### ISSUE TITLE
Adding A-B Loop indicator on progress bar

### ISSUE THREAD

**Issue by braineo** [iina/iina#548]
**System and IINA version:**

- macOS 10.12.4
- IINA 0.0.7

**Expected behavior:**

**Actual behavior:**

It will be intuitive if we can highlight the loop section on progress bar and provide handle for adjusting loop duration
1. Easy to know the looping status.
1. No need to clear loop and start all over again.

- Crash report:
```

```

- MPV log:
```

```

**Steps to reproduce:**

- [x] MPV does not have this problem.

**How often does this happen?**

**Comment by braineo** [comment]
Take a look at this. It seems we need a custom slider. NSSlider does not support this directly.

**Comment by low-batt** [comment]
Four years later and this excellent proposal from @braineo has still not been implemented. So I decided to take a stab at it. I will shortly submit a pull request for a proposed implementation for this feature. That is the good news.

The bad news is that I am not a graphical user interface programmer and have no talent for UI design, therefore my submission may be deemed as putrid by the core developers and unworthy of inclusion in IINA. Nevertheless I am blindly plunging head since no one else has had the time to implement this. Hopefully my submission will be considered adequate as an initial implementation with a more stylish UI slated as a future enhancement.

Needless to say a vicious brutal savage review is definitely called for. Everything is up for criticism. Have at it.

@braineo: Does the following line up with what you envisioned?:

### User Experience

When the user clicks on the `A-B Loop` command in the `Playback` menu or presses the Shift(⇧)-Command(⌘)-L key combination to set the A point an additional thumb appears on the progress bar's slider in front of the primary thumb:

![issue-548-a-light](https://user-images.githubusercontent.com/86170219/139542721-e5f8b515-2723-4360-9e8a-204e0221bd10.jpg)

![issue-548-a-dark](https://user-images.githubusercontent.com/86170219/139542726-11ba2ab6-5027-44ed-9c4f-2575952cf44e.jpg)

Hovering over the thumb displays a short help tag:

![issue-548-a-tooltip-light](https://user-images.githubusercontent.com/86170219/139542731-4bd34b3e-78f5-4560-9f3e-8b1508f754b0.jpg)

![issue-548-a-tooltip-dark](https://user-images.githubusercontent.com/86170219/139542745-8d847222-86b8-40dc-9758-ae8ef66fbea8.jpg)

When the user invokes the `A-B Loop` command a second time another thumb appears on the progress bar's slider:

![issue-548-b-tooltip-light](https://user-images.githubusercontent.com/86170219/139542804-c014a7e6-81fc-47ab-9dff-533251784436.jpg)

![issue-548-b-tooltip-dark](https://user-images.githubusercontent.com/86170219/139542818-47d8ee87-050c-4de2-8b64-42b095a83ef6.jpg)

Moving either thumb changes the associated loop point in [mpv](https://mpv.io/manual/stable/). Unlike the progress thumb, moving either of the loop thumbs does not pause play. While a loop thumb is being moved the onscreen display shows the position within the video:

![issue-548-b-osd-light](https://user-images.githubusercontent.com/86170219/139542759-71efdb19-2347-4aeb-b16f-84be9857faf5.jpg)

![issue-548-b-osd-dark](https://user-images.githubusercontent.com/86170219/139542764-1693a45e-cc73-4997-a6de-abc5794c0308.jpg)

When the user invokes the `A-B Loop` command for a third time the loop is cleared and both thumbs will disappear.

The state of the `A-B Loop` feature is stored by  [mpv](https://mpv.io/manual/stable/) in the watch later config files. If a video is played again loop thumbs will be displayed if the respective loop point was set when the video was last played.

Dragging a loop thumb requires carefully clicking directly on the thumb. If the cursor is not over the thumb then the slider's bar will be clicked which moves the progress thumb. Requiring exact clicking on the loop thumb is intentional in order to avoid accidental changing of the loop points.

For the same reason when thumbs overlap the thumb that is selected by a click is based on the following order:
- Progress thumb
- B loop point
- A loop point

The thumbs conform to the functionality provided by the [mpv](https://mpv.io/manual/stable/) `A-B Loop` feature which is very forgiving and allows the B point to be set before the A point. From the [mpv](https://mpv.io/manual/stable/) manual:

> **--ab-loop-a=\<time>, --ab-loop-b=\<time>**
>
>    Set loop points. If playback passes the b timestamp, it will seek to the a timestamp. Seeking past the b point doesn't loop (this is intentional).
>
>   If a is after b, the behavior is as if the points were given in the right order, and the player will seek to b after crossing through a. 

So the movement of the thumbs is unrestricted.

### Implementation

The commit will:

- Add a new class `PlaySliderLoopKnob` which adds a thumb to the slider to represent a loop point. This is a subclass of `NSView` to gain support for tool tips.

- Add new class `PlaySlider`, a subclass of `NSSlider`. This class creates two `PlaySliderLoopKnob` objects in order to add two thumbs to the progress slider. This class also has a side effect of suppressing needless drawing when the slider is hidden which spindumps reveal is currently happening. 

- Change `MainWindowController.xib` to use the custom class `PlaySlider` for the progress slider.

- Change `PlayerCore` to expose the A and B loop set points and allow them to be updated

- Change `MainWindowController` to observe the location of the loop knobs on the play slider and update the new properties exposed by `PlayerCore`. This class also establishes tracking areas to know when the mouse is moving these knobs.

- Change `PlayerWindowController` to synchronize the thumbs with the `mpv` loop points

- Change `MainMenuActions` to call the window controller for the `A-B Loop` command instead of `PlayerCore` to keep the thumbs synchronized with the `mpv` state

- Add a `LoopStatus` enum to `PlaybackInfo` to replace the use of integers for the value of  `abLoopStatus`

- Add a new message to `OSDMessage` for when the new thumbs are moved

Thoughts? Critical reviews including criticism of style and naming are welcome.

**Comment by braineo** [comment]
@low-batt WOW! I would say it actually goes beyond! Thank you!

I hope this can pass the review soon and get merged!

**Comment by Ralph-y** [comment]
Can I get this, like, now?  I'm still looking for that elusive QT7Pro replacement in my workflow (I maintain a virtual machine for it but it's clunky to use) and this fits perfectly.

**Comment by low-batt** [comment]
I'm merely an IINA user, not a part of the IINA development team. I've been investigating issues and proposing solutions to help out the development team. The changes for this issue are awaiting their review. The latest status from the development team is that the new plug-in system is nearing completion. Hopefully once that is done and merged they will look into the outstanding pull requests.

As you can see above CarterLi has merged these changes into his IINA fork along with his changes to add HDR support for M1 Pro/Max. He has published a release you could try.

**Comment by Ralph-y** [comment]
I'll check that out.  Thanks!

**Comment by Ralph-y** [comment]
Is this dead?  All IINA lacks is a-b loop handles to be the better-than-perfect QTP7 replacement.

**Comment by low-batt** [comment]
The last 3 commits you see above are me updating the proposed changes so that they will work with the latest changes that were added to IINA, keeping these changes ready to merge. The next step is for other developers to review, approve and merge these changes.

The holdup is that this is a new feature and fixing defects has been the priority. I'm expecting that fixing defects will still be a priority for the next release. I will be pushing the team to include this feature in that release. I can't say if it will make it in or be delayed until the next feature release, that is up to the other developers.

But definitely not dead!

**Comment by low-batt** [comment]
YES. This was one of my early PRs. I tried hard to not change the existing code. I remember it seemed to me positioning of the main slider knob seemed off. It was the knob going past the B knob that I noticed. And yes, that is easier to see in the floating OSC. I missed the other issues you noted. I was thinking any existing issues with the OSC should be handled in a separate PR.

With this PR I wanted to establish base code that was working. I was expecting an actual UI programer could then tune it up. Initially I was trying to use arrows for the A and B knobs. They looked awful due to the small size of the progress bar. I ended up with what you see. I did not try and implement some of the UI details, like the shadow the main knob has.

I've been a little slow to respond as I have been distracted by the need to shovel snow after a winter storm. I did notice the PR updates. I will try and test them out soon.

**Comment by svobs** [comment]
> I ended up with what you see. I did not try and implement some of the UI details, like the shadow the main knob has.

I think they look good in Dark Mode. In Light Mode I think they also look fine -especially when compared to nothing at all. Merging this does not preclude further style changes some point in the future so I don't think it's necessary to get snagged on that.

I don't see a shadow on the main knob...(?)

The more I think about it though, the more I think the issue of the main knob's X offset should be addressed. I did do a medium dive into the issue, and it's not trivial. It might be Apple which broke it. When I reverted to the default box-shaped knob, I could see why it is doing what it is: it's adjusting the slider position toward the ends so that the edge of the "box" touches the end of the slider "track" rather than the center of the box. I'll file an issue for this soon.

> I've been a little slow to respond as I have been distracted by the need to shovel snow after a winter storm.

I got a sense you might be located in the Midwest, heh. That is a joy I have yet to appreciate. Though the storms here were shockingly violent at times, frequent brief power outages aren't too bothersome when you've got a full laptop battery and decent cell service.

**Comment by low-batt** [comment]
I see I messed up and replied to the issue when I intended to reply to the PR. The code for drawing a shadow is [here](https://github.com/iina/iina/blob/da1e06cf41c32390356e2b69120cc969bcd31612/iina/PlaySliderCell.swift#L121) in `PlaySliderCell`.

I grew up in the Midwest. I'm currently in New England. Not frigid like the Midwest, but cold enough for a snowy Winter.
